core/src/ledger/storage/masp_conversions.rs

```diff
@@ -27,6 +39,143 @@ pub struct ConversionState {
+    addr: &Address,
+) -> crate::ledger::storage_api::Result<(u32, u32)>
+where
+    D: 'static + super::DB + for<'iter> super::DBIter<'iter>,
+    H: 'static + super::StorageHasher,
+{
+    let masp_addr = address::masp();
+    // Query the storage for information
+
+    //// information about the amount of tokens on the chain
+    let total_tokens: token::Amount = wl_storage
+        .read(&token::minted_balance_key(&wl_storage.storage.native_token))?
+        .expect("the total supply key should be here");
+
+    // total staked amount in the Shielded pool
+    let total_token_in_masp: token::Amount = wl_storage
+        .read(&token::balance_key(addr, &masp_addr))?
+        .unwrap_or_default();
+
+    let epochs_per_year: u64 = wl_storage
+        .read(&parameters::storage::get_epochs_per_year_key())?
+        .expect("");
+
+    //// Values from the last epoch
+    let last_inflation: token::Amount = wl_storage
+        .read(&token::masp_last_inflation(addr))
+        .expect("failure to read last inflation")
+        .expect("");
+
+    let last_locked_ratio: Dec = wl_storage
+        .read(&token::masp_last_locked_ratio(addr))
+        .expect("failure to read last inflation")
+        .expect("");
+
+    //// Parameters for each token
+    let max_reward_rate: Dec = wl_storage
+        .read(&token::masp_max_reward_rate(addr))
+        .expect("max reward should properly decode")
+        .expect("");
+
+    let kp_gain_nom: Dec = wl_storage
+        .read(&token::masp_kp_gain(addr))
+        .expect("kp_gain_nom reward should properly decode")
+        .expect("");
+
+    let kd_gain_nom: Dec = wl_storage
+        .read(&token::masp_kd_gain(addr))
+        .expect("kd_gain_nom reward should properly decode")
+        .expect("");
+
+    let locked_target_ratio: Dec = wl_storage
+        .read(&token::masp_locked_ratio_target(addr))?
+        .expect("");
+
+    // Creating the PD controller for handing out tokens
+    let controller = RewardsController {
+        locked_tokens: total_token_in_masp,
+        total_tokens,
+        locked_ratio_target: locked_target_ratio,
+        locked_ratio_last: last_locked_ratio,
+        max_reward_rate,
+        last_inflation_amount: last_inflation,
+        p_gain_nom: kp_gain_nom,
+        d_gain_nom: kd_gain_nom,
+        epochs_per_year,
+    };
+
+    let ValsToUpdate {
+        locked_ratio,
+        inflation,
+    } = RewardsController::run(controller);
+
+    // inflation-per-token = inflation / locked tokens = n/PRECISION
+    // ∴ n = (inflation * PRECISION) / locked tokens
+    // Since we must put the notes in a compatible format with the
+    // note format, we must make the inflation amount discrete.
+    let noterized_inflation = if total_token_in_masp.is_zero() {
+        0u32
+    } else {
+        crate::types::uint::Uint::try_into(
+            (inflation.raw_amount() * PRECISION)
+                / total_token_in_masp.raw_amount(),
+        )
+        .unwrap()
```

Comment:
Is this any different? Otherwise I don't see much about this which is different

Reply:
Other than use of unsigned amounts, this code is probably the same as what's in the other branches/PRs.